### PR TITLE
SCP-3315: Added a (named)debruijn scope-checking pass

### PR DESCRIPTION
**Pull request by bezirg**
Adds a scope-checking pass as a requirement before the transition to direct-debruijn evaluation.
This PR only adds an implementation and test code; it does not enable the pass.
The pass should be later enabled by placing it in ledger-api , after script flat-desirialization but *before* cek evaluation.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    Delay _ t     -> checkScope t
+    _             -> True
+  where
+    -- the current level as a reader value
+    withL :: Word -> UPLC.Term name uni fun a -> Bool
+    withL !lvl = \case
+        -- var should be larger than 0 (satisfied by Index=Natural/Word)
+        -- var should be LEQ to the current level
+        Var _ n       -> fromIntegral (n ^. index) <= lvl
+        LamAbs _ _ t  -> withL (lvl+1) t
```
Comment (michaelpj) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs:28]: Hmmm. What happened in the past when the index was wrong (i.e. not 0)? I see that [now](https://github.com/input-output-hk/plutus/blob/master/plutus-core/untyped-plutus-core/src/UntypedPlutusCore/DeBruijn.hs#L138) we ignore it and just act as if it was 0, but I think that's not what we did in the actual Alonzo HF, we actually treated the index as if it mattered (this is true, the behaviour changed [here](https://github.com/input-output-hk/plutus/commit/da4f85cdd2a3a261ce540e8dc51d2a3c5fa89ed2#diff-33151f0a151d48e22c6d4855933528c1c7fbace6dd470d218bc27ba38c25e121).

However, answering my own question: we omit indices on binders in the serialized form, and put them in as 0 when we deserialise, so we can never have wrong indices. Nonetheless, I think it's cheap and sensible to check that they're all 0 here. With a comment.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+module UntypedPlutusCore.Check.Scope (checkScope) where
+
+import Control.Lens hiding (index)
+import UntypedPlutusCore.Core.Type as UPLC
+import UntypedPlutusCore.DeBruijn as UPLC
+
+checkScope :: forall name uni fun a. HasIndex name
+           => UPLC.Term name uni fun a
+           -> Bool
+checkScope = \case
```
Comment (michaelpj) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs:13]: I looked at this and thought "can't you just replace this with `withL 0`? But then we'd allow a top level `Var 0`, which we don't want. 

Alternative: just use one function and add a clause to the `Var` case that says `&& index > 0`. I think I'd dislike that less.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+{-# LANGUAGE ScopedTypeVariables #-}
+module UntypedPlutusCore.Check.Scope (checkScope) where
+
+import Control.Lens hiding (index)
+import UntypedPlutusCore.Core.Type as UPLC
+import UntypedPlutusCore.DeBruijn as UPLC
+
+checkScope :: forall name uni fun a. HasIndex name
+           => UPLC.Term name uni fun a
+           -> Bool
```
Comment (michaelpj) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs:12]: I think it's worth returning a structured error here. Useful for users, easier to test.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs (hunk 2 above)
Comment (bezirg) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs:13]: the `&& index > 0` would makeit potentially slower (slightly?), and this pass we will run after every script deserialization.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs (hunk 1 above)
Comment (bezirg) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs:28]: fair enough.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs (hunk 2 above)
Comment (michaelpj) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs:13]: True, but in a pretty trivial way. We can do it this way and then benchmark it :) Doing performance optimizations without numbers is risky.

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+2) that all binders are set to debruijn index 0.
+
+Feeding the result of the debruijnification to this function is expected to pass.
+
+On the other hand, because of (2), this pass is
+stricter than the undebruijnification's (indirect)
+scope-checking, see NOTE: [DeBruijn indices of Binders].
+-}
+checkScope :: forall e m name uni fun a.
+             (HasIndex name, MonadError e m, AsFreeVariableError e)
```
Comment (michaelpj) [inline plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Check/Scope.hs:24]: I'd maybe have had an error just for this pass, but reusing that one seems fine.

**Review by michaelpj — APPROVED** [review]
LGTM, I haven't looked through the test cases individually but I trust that you have!

**Pull request merged**